src/components/tl/mlx5/mcast/tl_mlx5_mcast_allgather.c

```diff
@@ -270,6 +270,152 @@ void ucc_tl_mlx5_mcast_allgather_progress(ucc_coll_task_t *coll_task)
+
+/*
+ * at each stage half of the mcast groups are ready for receiving mcast
+ * packets while the other half are getting prepared by preposting recv
+ * buffers
+ */
+static inline ucc_status_t
+ucc_tl_mlx5_mcast_prepare_zero_copy_allgather(ucc_tl_mlx5_mcast_coll_comm_t *comm,
+                                              ucc_tl_mlx5_mcast_coll_req_t  *req)
+{
+    ucc_tl_mlx5_mcast_reg_t                   *reg    = NULL;
+    ucc_rank_t                                 root   = 0;
+    int                                        offset = 0;
+    ucc_status_t                               status;
+    ucc_rank_t                                 j, i;
+    int                                        total_steps;
+    ucc_tl_mlx5_mcast_pipelined_ag_schedule_t *new_sched;
+
+    ucc_assert(comm->allgather_comm.truly_zero_copy_allgather_enabled);
+
+    req->concurreny_level = comm->mcast_group_count / 2;
+    req->concurreny_level = ucc_min(req->concurreny_level, ONE_SIDED_MAX_CONCURRENT_LEVEL);
+    req->concurreny_level = ucc_min(req->concurreny_level, comm->commsize);
+
+    if (req->concurreny_level == 0) {
+        tl_warn(comm->lib, "not enough concurreny level to enable zcopy pipeline allgather");
+        return UCC_ERR_NOT_SUPPORTED;
+    }
+
+    if (comm->allgather_comm.mcast_prepost_bucket_size > req->num_packets) {
+        req->mcast_prepost_bucket_size = req->num_packets;
+    } else {
+        req->mcast_prepost_bucket_size = comm->allgather_comm.mcast_prepost_bucket_size;
+    }
+
+	status = ucc_tl_mlx5_mcast_validate_zero_copy_allgather_params(comm, req);
+	if (status != UCC_OK) {
+        return status;
+    }
+
+    /* calculate the schedule and details of what we should
+     * mcast and prepost to which mcast group at each stage*/
+    total_steps = req->num_packets * (comm->commsize / req->concurreny_level)
+                / req->mcast_prepost_bucket_size + 1;
+
+    new_sched = ucc_calloc(1, sizeof(ucc_tl_mlx5_mcast_pipelined_ag_schedule_t) * total_steps, "sched");
+    if (!new_sched) {
+        tl_warn(comm->lib, "cannot allocate memory for schedule list");
+        return UCC_ERR_NO_MEMORY;
+    }
+
+    /* generate schedule */
+    for (i = 0; i < total_steps; i++) {
+        ucc_assert(root < comm->commsize);
+        if (i < total_steps - 1) {
+            for (j = 0; j < req->concurreny_level; j++) {
+                new_sched[i].prepost_buf_op[j].group_id = j + req->concurreny_level * (i % 2);
+                new_sched[i].prepost_buf_op[j].offset   = offset * comm->max_per_packet;
+                new_sched[i].prepost_buf_op[j].root     = root + j;
+                new_sched[i].prepost_buf_op[j].count    = req->mcast_prepost_bucket_size;
+            }
+        } else {
+            new_sched[i].prepost_buf_op_done = 1;
+        }
+
+        if (i > 0) {
+            for (j = 0; j < req->concurreny_level; j++) {
+                new_sched[i].multicast_op[j].group_id     = new_sched[i - 1].prepost_buf_op[j].group_id;
```

Comment:
it seems that it way exceeds Col 80. Have you run `git-clang-format`?

@Sergei-Lebedev @janjust do we want this formatting style to be compulsory? If yes we should have the CI check for it.